rfcs/20201027-modular-tensorflow-graph-c-api.md

```diff
@@ -0,0 +1,527 @@
+Modular TensorFlow RFC designs a plugin architecture for several TensorFlow components(`Networking`, `Filesystems`, `Kernel`, `Graph` and `Accelerator backends`) through a stable ABI. This RFC describes the `Graph` module in the TensorFlow proper side, by introducing pluggable custom graph optimizer to the TensorFlow Grappler classes.
+The pluggable graph optimizer discovery and initialization is transparent to end users. As long as the graph plugin libraries follow the design described in this RFC, it can be plugged to TensorFlow proper and add a new graph optimizer into TensorFlow Grappler.
+
+Caveat: The proposed C API in its current form will not be compatible with the new TensorFlow runtime ([TFRT](https://blog.tensorflow.org/2020/04/tfrt-new-tensorflow-runtime.html)) and graph compiler.
+1. The nature and order of passes in the new compiler will differ significantly, probably allowing only limited reuse of algorithms or patterns from plugins developed for the existing runtime.
+2. The graph compiler will not communicate with the plugin using a GraphDef based format, but some TBD format, likely based on serialized [MLIR](https://www.tensorflow.org/mlir).
+
+
+## **User Benefit**
+
+This RFC provides a plugin infrastructure for TensorFlow to optimize graph with new custom optimizers, as long as users set up the system properly installing the graph plugin.
+
+## **Design Proposal**
+
+### Design Overview
+
+This RFC is intended to provide a new mechanism for custom graph optimizers, along with C APIs for users to implement and register their own plugable graph optimizers in Grappler.
+The C APIs follows current implementation C++ API, [TF_Buffer](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/c/c_api.h#L109-L113) and related proto files are the interface between proper and plugin.
+When initializing, TensorFlow loads the plugin and registers a new graph optimizer into Grappler. In the [Optimize](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/grappler/optimizers/graph_optimizer.h#L58) function, plugin authors need to deserialize `TF_Buffer` to `plugin::GraphDef` object to do some graph transformations, and serialize the optimized `plugin::GraphDef` object back to `TF_Buffer` as output.
+
+<p align="center">
+ <img src="20201027-modular-tensorflow-graph-c-api/flow.png" height="400"/>
+</p> 
+
+<!-- Here is what plugin authors need to do:
+1. Implementing `P_Optimize` function.
+2. Copying proto files from proper to plugin, implementing `P_MessageToBuffer`/`P_BufferToMessage` functions.
+3. (Optional)Implementing `plugin::util` functions for graph transformation.
+4. Registering the optimizer -->
+
+### Struct/Function/Object Overview
+- Struct
+  - Struct that should be filled by the plugin: `P_OptimizerConfigFns`, `P_RegistrationParams`, `TF_OptimizerBuilder`
```

Comment:
Does `P_` prefix mean that the struct should be filled by the plugin? I am thinking if `TP_` prefix would be better here (that might match what we do for `StreamExecutor` C API better where we have `SP` and `SE` prefixes).

Also, should `TF_OptimizerBuilder` be `(T)P_OptimizerBuilder`?

Reply:
Yes. I have already changed those names.